### PR TITLE
Geom and test utils for integrating obstacles

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Added/modified geom and test util functions to support integrating obstacles:
* distance for Circle
* stream operators for Circle and Polygon
* equalWithinTolerance
* new Obstacle hierarchy
And also some clean up

(This has been broken out from #1320 since that one was getting too big)

### Testing Done
Additional unit testing 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/motion_constraint/motion_constraint.h [hunk 1]
```diff
@@ -15,7 +15,7 @@ MAKE_ENUM(MotionConstraint,
           FRIENDLY_DEFENSE_AREA,
           // The box around the enemy goal
           ENEMY_DEFENSE_AREA,
-          // The box around the enemy goal, inflated by 0.2 meters for certain situations
+          // The box around the enemy goal, inflated by 0.3 meters for certain situations
```
Comment (garethellis0) [inline src/software/ai/motion_constraint/motion_constraint.h]: Where is `0.3` coming from here? It seems likely that this comment will quickly become out-of-date again.

File: src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    /**
+     * Gets the underlying circle
+     *
+     * @return circle
+     */
+    const Circle getCircle(void) const;
+
+   private:
+    Circle circle_;
```
Comment (garethellis0) [inline src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:34]: Perhaps we could have `CircleObstacle` extend both `Obstacle` and `Circle` so that we don't need to manually wrap functions?

File: src/software/ai/navigator/obstacle/new_obstacle/polygon_obstacle.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * Gets the underlying polygon
+     *
+     * @return polygon
+     */
+    const Polygon getPolygon(void) const;
+
+   private:
+    Polygon polygon_;
```
Comment (garethellis0) [inline src/software/ai/navigator/obstacle/new_obstacle/polygon_obstacle.h:33]: Similar comment here, why not have `PolygonObstacle` extend `Obstacle` and `Polygon`?

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp [hunk 4]
```diff
@@ -264,8 +264,8 @@ void GoalieTactic::calculateNextAction(ActionCoroutine::push_type &yield)
 
             // we want to restrict the block cone to the friendly crease, also potentially
             // scaled by a defense_area_deflation_parameter
-            Rectangle deflated_defense_area = field.friendlyDefenseArea();
-            deflated_defense_area.expand(-defense_area_deflation);
+            Rectangle deflated_defense_area =
+                *field.friendlyDefenseArea().expand(-defense_area_deflation);
```
Comment (garethellis0) [inline src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: Should we not check that `expand` didn't return`std::nullopt` here?

File: src/software/ai/hl/stp/tactic/goalie_tactic_test.cpp [hunk 5]
```diff
@@ -66,10 +66,8 @@ TEST_P(GoalieRestrainTest, goalie_position_safe)
     // test to make sure that points given inside of the rectangle
     // are not altered and are the same points
     // are constrained inside.
-    auto big_rectangle = world.field().friendlyDefenseArea();
-
     // blow up rectangle to a huge amount, to contain all the points
-    big_rectangle.expand(5);
+    auto big_rectangle = *world.field().friendlyDefenseArea().expand(5);
```
Comment (garethellis0) [inline src/software/ai/hl/stp/tactic/goalie_tactic_test.cpp]: Similar comment here, we should make sure we're checking optionals.

File: src/software/ai/navigator/obstacle/new_obstacle/obstacle.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+     *
+     * @return string that describes the obstacle
+     */
+    virtual std::string toString(void) const = 0;
+};
+
+/**
+ * We use a pointer to obstacles to support inheritance
+ */
+using ObstaclePtr = std::shared_ptr<Obstacle>;
```
Comment (garethellis0) [inline src/software/ai/navigator/obstacle/new_obstacle/obstacle.h:55]: Sorry, I'm not quite following why this is required?

File: src/software/new_geom/rectangle.cpp [hunk 7]
```diff
@@ -69,20 +69,20 @@ Point Rectangle::furthestCorner(const Point &p)
...
     }
 
     points_[0] = points_[0] + Vector(-amount, -amount);
     points_[1] = points_[1] + Vector(-amount, amount);
     points_[2] = points_[2] + Vector(amount, amount);
     points_[3] = points_[3] + Vector(amount, -amount);
 
-    return true;
+    return std::make_optional(*this);
 }
```
Comment (garethellis0) [inline src/software/new_geom/rectangle.cpp]: So this interface is a bit more confusing for me now. Generally I would expect something to have a return value or modify the state of the object, but not both. This is definitely worth having a larger discussion about, as we'll likely see this pattern again for geom stuff.

@jsn0716 @MathewMacDougall @akhilveeraghanta - your thoughts?

File: src/software/new_geom/rectangle.h [hunk 8]
```diff
@@ -101,10 +101,10 @@ class Rectangle : public ConvexPolygon
      * @param amount The amount to expand or shrink the rectangle by on all sides, can be
      * positive or negative
      *
-     * @return bool Whether it was possible to expand/shrink the rectangle by
-     * amount requested, rectangle remains unchanged if impossible to expand/shrink
+     * @return expanded rectangle if it was possible to expand/shrink the rectangle by
+     * amount requested, and return std::nullopt if it wasn't possible
```
Comment (garethellis0) [inline src/software/new_geom/rectangle.h]: If we do elect to keep this interface style (return value + side effects) we should definitely note the side effects in this jdoc.

File: src/software/test_util/test_util.cpp [hunk 9]
```diff
@@ -102,4 +102,45 @@ namespace Test
...
+        return true;
+    }
+
+    bool TestUtil::checkIfVisuallySimilar(const Point &point1, const Point &point2)
+    {
+        return (point1.distanceFromPoint(point2) <=
+                (sqrt(2) * METERS_PER_MILLIMETER + GeomConstants::FIXED_EPSILON));
+    }
+
+    bool checkIfVisuallySimilar(const Circle &circle1, const Circle &circle2) {}
```
Comment (garethellis0) [inline src/software/test_util/test_util.cpp]: So I think the concept of being able to check geometry with some specified tolerance is definitely a great idea. Here's a few tweaks we could make to help clarity a bit:
- `visuallySimilar` is very ambiguous, perhaps `checkGeometryEqualWithTolerance`?
- we should expose the tolerance as an argument in the functions, even if we assign it a default value, because there's lots of cases where we might want to change that, and I wouldn't want to presume a tolerance for every use case here.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good, got a few suggestions on how we can reduce the number of wrapper functions by changing the inheritance a bit, but nothing major beyond that.

File: src/software/ai/navigator/obstacle/new_obstacle/obstacle.h (hunk 6 above)
Comment (jonathanlew) [reply src/software/ai/navigator/obstacle/new_obstacle/obstacle.h:55]: the idea is that we need to use pointers to the virtual class to use inheritance so ObstaclePtr is a shortcut

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp (hunk 4 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: yeah that's a detail that was missing in the existing code. It really should've checked for expand returning false

File: src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:34]: looked into it and it's difficult to support inheritance->can't think of a good way to write getCircle(). Generally, I think composition is a better design pattern than inheritance

File: src/software/ai/navigator/obstacle/new_obstacle/obstacle.h (hunk 6 above)
Comment (garethellis0) [reply src/software/ai/navigator/obstacle/new_obstacle/obstacle.h:55]: Ahhh I see. Perhaps we could change this to something like `convenience typedef` instead? I was just a bit confused with the current phrasing because I tried to read too much into it.

File: src/software/ai/motion_constraint/motion_constraint.h (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/motion_constraint/motion_constraint.h]: yeah thoughts on making this a dynamic param?

File: src/software/test_util/test_util.cpp (hunk 9 above)
Comment (jonathanlew) [reply src/software/test_util/test_util.cpp]: ok changed it to that

File: src/software/ai/navigator/obstacle/new_obstacle/obstacle.h (hunk 6 above)
Comment (jonathanlew) [reply src/software/ai/navigator/obstacle/new_obstacle/obstacle.h:55]: fixed it

File: src/software/new_geom/rectangle.h [hunk 10]
```diff
@@ -101,10 +101,10 @@ class Rectangle : public ConvexPolygon
...
      */
-    bool expand(double amount);
+    std::optional<Rectangle> expand(double amount);
```
Comment (MathewMacDougall) [inline src/software/new_geom/rectangle.h]: nit: if this function can expand or shrink, can we rename it to `scale` or something? Right now it's not obvious it can be used for shrinking.

File: src/software/ai/motion_constraint/motion_constraint.h (hunk 1 above)
Comment (MathewMacDougall) [reply src/software/ai/motion_constraint/motion_constraint.h]: Maybe we shouldn't comment the values of expansion / modification in the enum definition? The inflation amount is more of an implementation detail.

File: src/software/new_geom/rectangle.cpp (hunk 7 above)
Comment (MathewMacDougall) [reply src/software/new_geom/rectangle.cpp]: I agree that both at once is confusing. I don't have a super strong personal preference as long as we're consistent. We already follow the pattern of having a return value and not modifying object state, so I'd probably vote for just sticking with that convention.

File: src/software/test_util/test_util.cpp [hunk 11]
```diff
@@ -102,4 +102,50 @@ namespace Test
         return Robot(robot_id_counter++, pt, Vector(), Angle(), AngularVelocity(),
                      Timestamp());
     }
+
+    bool TestUtil::checkGeometryEqualWithTolerance(const Polygon &poly1,
```
Comment (MathewMacDougall) [inline src/software/test_util/test_util.cpp]: Why not call this `checkPolygonEqual...`?

File: src/software/test_util/test_util.cpp [hunk 12]
```diff
@@ -102,4 +102,50 @@ namespace Test
...
+            if (!checkGeometryEqualWithTolerance(ppoints1[i], ppoints2[i], tolerance))
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    bool TestUtil::checkGeometryEqualWithTolerance(const Circle &circle1,
```
Comment (MathewMacDougall) [inline src/software/test_util/test_util.cpp]: same as above but for `Circle...`

File: src/software/test_util/test_util.cpp [hunk 13]
```diff
@@ -102,4 +102,50 @@ namespace Test
...
+    }
+
+    bool TestUtil::checkGeometryEqualWithTolerance(const Point &point1,
+                                                   const Point &point2, double tolerance)
+    {
+        return (point1.distanceFromPoint(point2) <=
+                (sqrt(2) * tolerance + GeomConstants::FIXED_EPSILON));
+    }
+
+    bool checkGeometryEqualWithTolerance(const Circle &circle1, const Circle &circle2) {}
```
Comment (MathewMacDougall) [inline src/software/test_util/test_util.cpp]: unused implementation, this is already defined above.

File: src/software/test_util/test_util.cpp [hunk 14]
```diff
@@ -102,4 +102,50 @@ namespace Test
...
+            return false;
+        }
+        if (fabs(circle1.getRadius() - circle2.getRadius()) > tolerance)
+        {
+            return false;
+        }
+        return true;
+    }
+
+    bool TestUtil::checkGeometryEqualWithTolerance(const Point &point1,
```
Comment (MathewMacDougall) [inline src/software/test_util/test_util.cpp]: same as above with `Point...`

File: src/software/test_util/test_util.cpp [hunk 15]
```diff
@@ -102,4 +102,50 @@ namespace Test
...
+                                                   const Point &point2, double tolerance)
+    {
+        return (point1.distanceFromPoint(point2) <=
+                (sqrt(2) * tolerance + GeomConstants::FIXED_EPSILON));
```
Comment (MathewMacDougall) [inline src/software/test_util/test_util.cpp]: Where's the `sqrt(2)` coming from?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:34]: https://en.wikipedia.org/wiki/Composition_over_inheritance

File: src/software/new_geom/rectangle.cpp (hunk 7 above)
Comment (jonathanlew) [reply src/software/new_geom/rectangle.cpp]: changed it to return a modified rectangle and I also refactored the failed shrinking case to a an optional bool& argument because we usually don't care if the expansion succeeded or not (we can just assume it succeeds)

File: src/software/new_geom/rectangle.h (hunk 10 above)
Comment (jonathanlew) [reply src/software/new_geom/rectangle.h]: changed to scale

File: src/software/new_geom/rectangle.h (hunk 8 above)
Comment (jonathanlew) [reply src/software/new_geom/rectangle.h]: see comment above

File: src/software/test_util/test_util.cpp (hunk 11 above)
Comment (jonathanlew) [reply src/software/test_util/test_util.cpp]: check___ is actually redundant so changing to equalWithinTolerance

File: src/software/test_util/test_util.cpp (hunk 15 above)
Comment (jonathanlew) [reply src/software/test_util/test_util.cpp]: yeah it's a bit arbitrary, I changed it to compare each difference in x/y value to the tolerance instead

File: src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h (hunk 2 above)
Comment (garethellis0) [reply src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:34]: Sorry, just to be clear here, what would the difficulty of supporting inheritance be here? At a quick glance it seems like we would be able to just make it extend `Circle` and then delete the extra wrapper functions we have?

I totally take your point about composition over inheritance, but it really seems like we're just wrapping functions here that we shouldn't need to.

File: src/software/new_geom/rectangle.cpp (hunk 7 above)
Comment (MathewMacDougall) [reply src/software/new_geom/rectangle.cpp]: I think we may want to stick with returning an optional in all cases though (or making the failure status always available). Right now this assumes we're basically using well-known / constant values as the scaling amount, but this assumption doesn't hold in all cases. Perhaps at some point we dynamically scale based on velocity or some random parameter that could make the scaling positive or negative. In this case we don't know exactly if we will be shrinking or expanding, so would want to check the failure. I think it's easier if we just have one function so developers don't have to be so aware of cases like this.

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp [hunk 16]
```diff
@@ -264,8 +264,8 @@ void GoalieTactic::calculateNextAction(ActionCoroutine::push_type &yield)
 
             // we want to restrict the block cone to the friendly crease, also potentially
             // scaled by a defense_area_deflation_parameter
-            Rectangle deflated_defense_area = field.friendlyDefenseArea();
-            deflated_defense_area.expand(-defense_area_deflation);
+            Rectangle deflated_defense_area =
+                field.friendlyDefenseArea().scale(-defense_area_deflation);
```
Comment (garethellis0) [inline src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: I could be mistaken, but I'm pretty sure `scale` multiplies the area instead of adding to it, so this is very much not the same?

File: src/software/test_util/test_util.cpp [hunk 17]
```diff
@@ -102,4 +102,37 @@ namespace Test
...
+                                        double tolerance)
+    {
+        return equalWithinTolerance(c1.getOrigin(), c2.getOrigin(), tolerance) &&
+               equalWithinTolerance(c1.getRadius(), c2.getRadius(), tolerance);
+    }
+
+    bool TestUtil::equalWithinTolerance(const Point &pt1, const Point &pt2,
+                                        double tolerance)
+    {
+        return equalWithinTolerance(pt1.x(), pt2.x(), tolerance) &&
```
Comment (MathewMacDougall) [inline src/software/test_util/test_util.cpp]: I think what you have before with the euclidean distance was more intuitive (what users would expect from comparing points with tolerance), just without the `sqrt(2)`

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/new_geom/rectangle.cpp (hunk 7 above)
Comment (garethellis0) [reply src/software/new_geom/rectangle.cpp]: Yeah, I generally agree. I think that:
- If there is an unsafe version that's easier/simpler to use, people will just always use that
- If we're fairly sure the operation is safe, we could just de-reference the returned optional without checking it (which we should really _never_ do)

File: src/software/test_util/test_util.cpp (hunk 17 above)
Comment (jonathanlew) [reply src/software/test_util/test_util.cpp]: ok changed that

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp (hunk 16 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: I believe it's additive

File: src/software/new_geom/rectangle.cpp (hunk 7 above)
Comment (jonathanlew) [reply src/software/new_geom/rectangle.cpp]: After analysing where scale/expand is being used, I've concluded that most of these situations would benefit from specialized functions. For example:
* Field::getInflatedDefenseArea and Field::getInflatedDefenseArea -> we should maintain one side of the rectangle on the goal line rather than expanding from the centre of the rectangle
* ObstacleFactory implementing specialized functions for creating expanded rectangles according to robot geometry and constants -> we should include ideas of infinite x/y sides to the rectangles and only expanding some sides of the rectangle so the obstacle reflects the real intention of the motion constraint 

In other words, I think that expand/scale is being used in too many different situations that it's hard to make it optimal in any situation


I think that this work is out of scope for integrating obstacles and should be tackled alongside #980
@MathewMacDougall @garethellis0 thoughts on reverting scale/expand changes and postponing them until we tackle how to create these geometries?

Comment (MathewMacDougall) [reply src/software/new_geom/rectangle.cpp]: I agree that the special cases you highlighted are currently not good uses of `expand`. I'm fine reverting and deferring for now, but can you please make tickets for
* Specifically fixing the `expand` function
* Fixing the special cases you listed

Comment (jonathanlew) [reply src/software/new_geom/rectangle.cpp]: motionconstraint obstacles: #1331
new Field functions: #1332
We can reassess how to approach Rectangle::expand after we fix these special cases

Comment (MathewMacDougall) [reply src/software/new_geom/rectangle.cpp]: Can we make a ticket for re-assessing `expand` so we don't forget?

Comment (jonathanlew) [reply src/software/new_geom/rectangle.cpp]: I don't see any point in making a ticket that doesn't have any specific goals or actions. I'm sure we'll remember to do it

File: src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:34]: Extending `Circle` would only give us `contains` and we would need to construct a representation of the superclass (can't think of a "nice" way to do this) to implement distance and intersects, since they use the geom util functions. The benefit here is one function at the cost of a lot more complexity.

File: src/software/ai/motion_constraint/motion_constraint.h (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/motion_constraint/motion_constraint.h]: removed the 0.2 meters part

File: src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h (hunk 2 above)
Comment (garethellis0) [reply src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:34]: Sorry, I think I'm misunderstanding something here, so just bear with me. I changed `CircleObstacle` to extend `public Circle`, and then wrote (and ran) the following test:
```
TEST(NavigatorObstacleTest, circle_distance){
    Circle expected({2, 2}, 3);
    CircleObstacle circle_obstacle(expected);

    ::distance(circle_obstacle, Point(0,0));
}
```
and it worked fine. The same thing should apply to the other functions we've wrapped here. 

I'm happy to call about this if you think it would be easier/faster to discuss that way!

**Comment by garethellis0** [comment]
Looking good, just the one comment left on inheritance for geometry stuff and then we should be good to go.

File: src/software/new_geom/rectangle.cpp (hunk 7 above)
Comment (MathewMacDougall) [reply src/software/new_geom/rectangle.cpp]: > I'm sure we'll remember to do it

I don't agree. We've had "investigation" tickets before. It sounds like we might want to do _something_, so I think it's fine for the goal of the ticket to be to define what action we want to take, if any.

Comment (jonathanlew) [reply src/software/new_geom/rectangle.cpp]: I've added an extra task for the assignee of #1332 to come up with a new ticket for cleaning up Rectangle::expand, since the assignee would probably know best

File: src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:34]: Ah I totally misunderstood what you meant. I didn't realize that you wanted to change the interface so that the caller directly calls util functions. That makes a lot more sense

Comment (jonathanlew) [reply src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:34]: @garethellis0 I remember why I designed it this way now. The goal is to have contains, distance, and intersects to work on `Obstacle`s without needing to know what type (Circle/Polygon) the Obstacle is. We would need a function like ::distance(Obstacle o, Point p), which is difficult to implement without knowing the underlying shape of obstacle, hence the inheritance structure we have here. Happy to call if you think that's easier

Comment (garethellis0) [reply src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:34]: Ok, that's reasonable, so I'm good with this as-is.

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

File: src/software/ai/motion_constraint/motion_constraint.h [hunk 18]
```diff
@@ -15,7 +15,7 @@ MAKE_ENUM(MotionConstraint,
           FRIENDLY_DEFENSE_AREA,
           // The box around the enemy goal
           ENEMY_DEFENSE_AREA,
-          // The box around the enemy goal, inflated by 0.2 meters for certain situations
+          // The box around the enemy goal, inflated for certain situations
```
Comment (MathewMacDougall) [inline src/software/ai/motion_constraint/motion_constraint.h]: super nit: Isn't there just a single size for the `INFLATED_ENEMY_DEFENSE_AREA`, it's just larger than the regular defense area by definition (hence the name inflated...)? Or does how much it's inflated actually change?

"inflated for certain situations" makes me think the size changes dynamically, which IIRC isn't true.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/software/ai/motion_constraint/motion_constraint.h (hunk 18 above)
Comment (jonathanlew) [reply src/software/ai/motion_constraint/motion_constraint.h]: yes there is a single size for INFLATED_ENEMY_DEFENSE_AREA, but that can change either by being manually tuned or potentially with a dynamic param, so it's better that we don't hard code it in this comment since the comment is likely to go out of date (it's already out of date)

Comment (MathewMacDougall) [reply src/software/ai/motion_constraint/motion_constraint.h]: Yes I agree with not hardcoding the value in the comment. What I meant to say was that the "inflated for certain situations" part makes it sound like it can change at runtime under certain cases (not from param tuning). I think the comment would be fine / more clear without the last part. Based on the name it's already obvious its an inflated / larger rectangle.

Maybe just "An inflated box around the enemy defense area"?

Comment (jonathanlew) [reply src/software/ai/motion_constraint/motion_constraint.h]: done

File: src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * An Obstacle represented by a Circle
+ */
+class CircleObstacle : public Obstacle
+{
+   public:
+    CircleObstacle() = delete;
+
+    /**
+     * Construct a circle obstacle by with a Circle
```
Comment (oswinso) [inline src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h]: nit: typo

File: src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+     *
+     * @param circle Circle to make obstacle with
+     */
+    explicit CircleObstacle(const Circle circle);
+
+    bool contains(const Point& p) const override;
+    double minDistance(const Point& p) const override;
+    bool intersects(const Segment& segment) const override;
+    std::string toString(void) const override;
```
Comment (oswinso) [inline src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:24]: I'm not sure what conventions your codebase follows, but having the `void` parameter is equivalent to not having it there.

**Review by oswinso — COMMENTED** [review]
LGTM!

File: src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h (hunk 20 above)
Comment (jonathanlew) [reply src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:24]: our codebase is a mix of C and C++ and the `void` argument does make a difference for C, so for simplicity we prefer having the `void` argument everywhere

Comment (oswinso) [reply src/software/ai/navigator/obstacle/new_obstacle/circle_obstacle.h:24]: I see, that makes sense.

**Review by garethellis0 — APPROVED** [review]

**Review by MathewMacDougall — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]
